File: solutions/scala/zebra-puzzle/1/src/main/scala/PersonFactory.scala
object PersonFactory {

    def createAll(): Iterator[Person] =
        createRange()
            .flatMap(createForColor)
            .map(identity)

    private def createForColor(color: Int): Iterator[Person] =
        createRange()
            .flatMap(nationality => createForNationality(color, nationality))

    private def createForNationality(color: Int, nationality: Int): Iterator[Person] =
        createRange()
            .flatMap(pet => createFor(color, nationality, pet))

    private def createFor(color: Int, nationality: Int, pet: Int): Iterator[Person] =
        createRange()
            .flatMap(drink => createForDrink(color, nationality, pet, drink))

    private def createForDrink(
                                  color: Int,
                                  nationality: Int,
                                  pet: Int,
                                  drink: Int
                              ): Iterator[Person] =
        createRange()
            .flatMap(hobby =>
                createHousePermutations(color, nationality, pet, drink, hobby)
            )

    private def createHousePermutations(
                                           color: Int,
                                           nationality: Int,
                                           pet: Int,
                                           drink: Int,
                                           hobby: Int
                                       ): Iterator[Person] =
        createRange()
            .map(house =>
                new Person(
                    Color.fromOrdinal(color),
                    Nationality.fromOrdinal(nationality),
                    Pet.fromOrdinal(pet),
                    Drink.fromOrdinal(drink),
                    Hobby.fromOrdinal(hobby),
                    House.fromOrdinal(house)
                )
            )

    private def createRange(): Iterator[Int] =
        Iterator.range(0, 5)
}

File: solutions/java/zebra-puzzle/1/src/main/java/PersonFactory.java
import java.util.stream.Stream;

final class PersonFactory {

    static Stream<Person> createAll() {
        return Stream.of(Color.values())
                     .flatMap(PersonFactory::createForColor);
    }

    private static Stream<Person> createForColor(Color color) {
        return Stream.of(Nationality.values())
                     .flatMap(nationality -> createForNationality(color, nationality));
    }

    private static Stream<Person> createForNationality(Color color, Nationality nationality) {
        return Stream.of(Pet.values())
                     .flatMap(pet -> createFor(color, nationality, pet));
    }

    private static Stream<Person> createFor(Color color, Nationality nationality, Pet pet) {
        return Stream.of(Drink.values())
                     .flatMap(drink -> createForDrink(color, nationality, pet, drink));
    }

    private static Stream<Person> createForDrink(
            Color color,
            Nationality nationality,
            Pet pet,
            Drink drink
    ) {
        return Stream.of(Hobby.values())
                     .flatMap(hobby -> createHousePermutations(color, nationality, pet, drink, hobby));
    }

    private static Stream<Person> createHousePermutations(
            Color color,
            Nationality nationality,
            Pet pet,
            Drink drink,
            Hobby hobby
    ) {
        return Stream.of(House.values())
                     .map(house -> new Person(
                             color,
                             nationality,
                             pet,
                             drink,
                             hobby,
                             house
                     ));
    }
}